File: debug-mcp-vscode/src/debugger/BreakpointSync.ts
import * as vscode from "vscode";
import { AgentBreakpoint, BridgeMessage, MessageTypes } from "../bridge/MessageProtocol";
import { BridgeClient } from "../bridge/BridgeClient";

export class BreakpointSync {
  private byAgentId = new Map<string, vscode.SourceBreakpoint>();

  constructor(private readonly bridge: BridgeClient) {}

  async handle(message: BridgeMessage) {
    if (message.type === MessageTypes.AgentSetBreakpoint && message.breakpoint) {
      await this.addAgentBreakpoint(message.breakpoint);
    }
    if (message.type === MessageTypes.AgentRemoveBreakpoint && message.breakpointId) {
      this.removeAgentBreakpoint(message.breakpointId);
    }
  }

  watch(context: vscode.ExtensionContext) {
    context.subscriptions.push(
      vscode.debug.onDidChangeBreakpoints((event) => {
        for (const removed of event.removed) {
          const agentId = [...this.byAgentId.entries()].find(([, bp]) => bp === removed)?.[0];
          if (agentId) {
            this.byAgentId.delete(agentId);
            this.bridge.send({
              type: MessageTypes.IdeBreakpointRemoved,
              breakpointId: agentId
            });
          }
        }
      })
    );
  }

  private async addAgentBreakpoint(breakpoint: AgentBreakpoint) {
    const uri = vscode.Uri.file(breakpoint.file);
    const location = new vscode.Location(
      uri,
      new vscode.Position(Math.max(0, breakpoint.line - 1), Math.max(0, (breakpoint.column ?? 1) - 1))
    );
    const sourceBreakpoint = new vscode.SourceBreakpoint(
      location,
      true,
      breakpoint.condition,
      undefined,
      undefined
    );
    vscode.debug.addBreakpoints([sourceBreakpoint]);
    this.byAgentId.set(breakpoint.id, sourceBreakpoint);
    this.bridge.send({
      type: MessageTypes.IdeBreakpointAdded,
      breakpointId: breakpoint.id,
      breakpoint
    });
  }

  private removeAgentBreakpoint(agentId: string) {
    const breakpoint = this.byAgentId.get(agentId);
    if (!breakpoint) return;
    vscode.debug.removeBreakpoints([breakpoint]);
    this.byAgentId.delete(agentId);
  }
}

File: debug-mcp-idea/src/main/kotlin/debugger/BreakpointSync.kt
package debugger

import bridge.AgentBreakpoint
import bridge.BridgeClient
import bridge.BridgeMessage
import bridge.MessageTypes
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.xdebugger.XDebuggerManager
import com.intellij.xdebugger.breakpoints.XBreakpoint
import com.intellij.xdebugger.breakpoints.XLineBreakpointType

class BreakpointSync(
    private val project: Project,
    private val bridge: BridgeClient
) {
    private val byAgentId = mutableMapOf<String, XBreakpoint<*>>()

    fun handle(message: BridgeMessage) {
        when (message.type) {
            MessageTypes.AgentSetBreakpoint -> message.breakpoint?.let { addAgentBreakpoint(it) }
            MessageTypes.AgentRemoveBreakpoint -> message.breakpointId?.let { removeAgentBreakpoint(it) }
        }
    }

    private fun addAgentBreakpoint(breakpoint: AgentBreakpoint) {
        val file = LocalFileSystem.getInstance().findFileByPath(breakpoint.file) ?: return
        val manager = XDebuggerManager.getInstance(project).breakpointManager
        val type = XLineBreakpointType.EXTENSION_POINT_NAME.extensionList.firstOrNull() ?: return
        val lineBreakpoint = manager.addLineBreakpoint(type, file.url, breakpoint.line - 1, null)
        byAgentId[breakpoint.id] = lineBreakpoint
        bridge.send(
            BridgeMessage(
                type = MessageTypes.IdeBreakpointAdded,
                breakpointId = breakpoint.id,
                breakpoint = breakpoint
            )
        )
    }

    private fun removeAgentBreakpoint(agentId: String) {
        val breakpoint = byAgentId.remove(agentId) ?: return
        XDebuggerManager.getInstance(project).breakpointManager.removeBreakpoint(breakpoint)
    }
}
